### PR TITLE
Collecting prefixes to exclude from pods and services

### PR DESCRIPTION
**Pull request by lobkovilya**
Signed-off-by: Ilya Lobkov <lobkovilya@yandex.ru>
## Description 
In case of absent of *kubeadm-config* nsmd-k8s starts monitor, that uses `Watch` method to be informed about new nodes or services. Since we know that there is one subnet for all Pods (*PodCIDR*) and one subnet for all Services (*ServiceCIDR*) we are able to make assumption about these subnets. Simply put we always find *the narrowest* subnet (subnet with longest prefix) that contains all service's IP addresses (or node's podCIDR). 

*Note:* it works pretty fine on real scenarios, with 3-4 services deployed, monitor usually predict correct ServiceCIDR
## Motivation and Context

Issue #839

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [X] Added integration testing to cover
- [X] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 1]
```diff
@@ -435,6 +436,43 @@ func fromIP(ip net.IP) (*big.Int, int) {
 	return val, 128
 }
 
+func min(a, b int) int {
```
Comment (nickolaev) [inline controlplane/pkg/prefix_pool/prefixpool.go]: Should this be part of our `tool` or `utils`? Looks pretty generic. Or maybe there is a package that already provides it?

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 2]
```diff
@@ -425,7 +425,8 @@ func toIP(ipInt *big.Int, bits int) net.IP {
 	}
 	return net.IP(ret)
 }
-func fromIP(ip net.IP) (*big.Int, int) {
+
+func FromIP(ip net.IP) (*big.Int, int) {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply controlplane/pkg/prefix_pool/prefixpool.go]: You are right, will fix

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 3]
```diff
@@ -402,20 +402,20 @@ func addressCount(pr string) uint64 {
...
 func clearNetIndexInIP(ip net.IP, prefixLen int) net.IP {
-	ipInt, totalBits := fromIP(ip)
+	ipInt, totalBits := FromIP(ip)
 	ipInt.SetBit(ipInt, totalBits-prefixLen, 0)
-	return toIP(ipInt, totalBits)
+	return ToIP(ipInt, totalBits)
 }
 
-func toIP(ipInt *big.Int, bits int) net.IP {
+func ToIP(ipInt *big.Int, bits int) net.IP {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply controlplane/pkg/prefix_pool/prefixpool.go]: Will fix

File: controlplane/pkg/nsmd/nsmd.go [hunk 4]
```diff
@@ -469,3 +464,56 @@ func getExcludedPrefixesFromConfigMap(serviceRegistry serviceregistry.ServiceReg
...
+	poolCh := make(chan prefix_pool.PrefixPool, 10)
+	var podSubnet string
+	var serviceSubnet string
+
+	go func() {
+		for {
+			extendResponse, err := stream.Recv()
+			if err != nil {
+				logrus.Error(err)
+				close(poolCh)
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply controlplane/pkg/nsmd/nsmd.go]: I think nsmd-k8s could accidentally fall down, so we don't have to terminate NSMD. But probably I need to provide some logic that will try to restore that stream

File: controlplane/pkg/nsmd/nsmd.go [hunk 5]
```diff
@@ -469,3 +464,56 @@ func getExcludedPrefixesFromConfigMap(serviceRegistry serviceregistry.ServiceReg
...
+				podSubnet = extendResponse.Subnet
+			}
+
+			if extendResponse.Type == registry.SubnetExtendingResponse_SERVICE {
+				serviceSubnet = extendResponse.Subnet
+			}
+
+			pool, err := prefix_pool.NewPrefixPool(append(additionalPrefixes, podSubnet, serviceSubnet)...)
+			if err != nil {
+				logrus.Error(err)
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply controlplane/pkg/nsmd/nsmd.go]: `NewPrefixPool` may return error in case of passed prefixes are not valid, so I suppose logging and continue is fine

File: controlplane/pkg/prefix_pool/prefixpool.go (hunk 1 above)
Comment (lobkovilya) [reply controlplane/pkg/prefix_pool/prefixpool.go]: That's funny but I found only `math.Min` that accepts `float` :) Moving this func to specific package is good idea

**Comment by edwarnicke** [comment]
@lobkovilya This appears to need a manual rebase.

File: test/kube_testing/rbac/cluster_roles.go [hunk 6]
```diff
@@ -114,6 +114,11 @@ func CreateRoleAdmin(name string, namespace string) Role {
 					Resources: []string{"configmaps"},
 					Verbs:     []string{"get"},
 				},
+				{
```
Comment (nickolaev) [inline test/kube_testing/rbac/cluster_roles.go]: That is a good finding. Can you please split it to a separate PR.

File: test/integration/basic_excluded_prefixes_test.go [hunk 7]
```diff
@@ -42,12 +42,14 @@ func TestExcludePrefixCheck(t *testing.T) {
...
 
+	defer nsmd_test_utils.FailLogger(k8s, nodes, t)
+
 	icmp := nsmd_test_utils.DeployICMP(k8s, nodes[0].Node, "icmp-responder-nse-1", defaultTimeout)
 
 	clientset, err := k8s.GetClientSet()
 	Expect(err).To(BeNil())
 
-	_, err = clientset.CoreV1().Pods(k8s.GetK8sNamespace()).Create(pods.NSCPod("nsc", nodes[0].Node,
+	_, err = clientset.CoreV1().Pods("default").Create(pods.NSCPod("nsc", nodes[0].Node,
```
Comment (nickolaev) [inline test/integration/basic_excluded_prefixes_test.go]: Hmm not sure? Why go back to default?

File: k8s/pkg/prefixcollector/collector_server_test.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+		"10.20.2.0/24",
+	}
+	expectedSubnets := []string{
+		"10.20.1.0/24",
+		"10.20.0.0/22",
+	}
+	checkSubnetWatcher(t, subnetSequence, expectedSubnets)
+}
+
+func TestLastIpsAlreadyInSubnet(t *testing.T) {
```
Comment (nickolaev) [inline k8s/pkg/prefixcollector/collector_server_test.go]: Please keep the tests in a seaprate file.

File: test/integration/basic_excluded_prefixes_test.go (hunk 7 above)
Comment (rdimitrov) [inline test/integration/basic_excluded_prefixes_test.go]: I think we shouldn't use the "default" namespace anymore 👍

File: test/kube_testing/rbac/cluster_roles.go (hunk 6 above)
Comment (nickolaev) [reply test/kube_testing/rbac/cluster_roles.go]: Oh actually maybe no. I let you decide.

File: k8s/pkg/prefixcollector/collector_server.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+			return nil, err
+		}
+		return ipNet, nil
+	}
+
+	return watchSubnet(nodeWatcher, keyFunc, subnetFunc)
+}
+
+func WatchServiceIpAddr(clientset *kubernetes.Clientset) (*SubnetWatcher, error) {
+	serviceWatcher, err := clientset.CoreV1().Services("default").Watch(metav1.ListOptions{})
```
Comment (rdimitrov) [inline k8s/pkg/prefixcollector/collector_server.go]: Same as the previous one about using the "default" namespace 👍

File: test/integration/basic_excluded_prefixes_test.go (hunk 7 above)
Comment (lobkovilya) [reply test/integration/basic_excluded_prefixes_test.go]: I haven't changed it intentionally, seems like consequences of rebase. Will fix

File: test/kube_testing/rbac/cluster_roles.go (hunk 6 above)
Comment (lobkovilya) [reply test/kube_testing/rbac/cluster_roles.go]: Tests on my branch will fail without these changes

File: k8s/pkg/prefixcollector/collector_server_test.go (hunk 8 above)
Comment (lobkovilya) [reply k8s/pkg/prefixcollector/collector_server_test.go]: What do you mean by separate file? Should I split `dummyWatcher` implementation and tests?

File: k8s/pkg/prefixcollector/collector_server.go (hunk 9 above)
Comment (lobkovilya) [reply k8s/pkg/prefixcollector/collector_server.go]: Thank you! Will fix

File: k8s/pkg/prefixcollector/collector_server_test.go (hunk 8 above)
Comment (nickolaev) [reply k8s/pkg/prefixcollector/collector_server_test.go]: It is already a test file, no need to change anything.
